pkg/multicluster/watcher.go

```diff
@@ -0,0 +1,193 @@
+package multicluster
+
+import (
+	"context"
+	"sync"
+
+	"github.com/rotisserie/eris"
+	"github.com/solo-io/go-utils/contextutils"
+	"github.com/solo-io/skv2/pkg/generated/kubernetes/core/v1/controller"
+	"github.com/solo-io/skv2/pkg/multicluster/kubeconfig"
+	"github.com/solo-io/skv2/pkg/reconcile"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	LocalCluster = ""
+)
+
+type ClusterWatcher interface {
+	// Run starts a watch for KubeConfig secrets on the cluster managed by the local manager.Manager.
+	// Note that Run will call Start on the local manager and run all registered ClusterHandlers.
+	Run(local manager.Manager) error
+	// RegisterClusterHandler adds a ClusterHandler to the ClusterWatcher.
+	RegisterClusterHandler(handler ClusterHandler)
+	// Stop cancels the context on every manager.Manager.
+	Stop()
+}
+
+type clusterWatcher struct {
+	ctx      context.Context
+	mutex    sync.RWMutex
+	handlers *handlerList
+	cancels  *cancelSet
+}
+
+var _ ClusterWatcher = &clusterWatcher{}
+
+func NewClusterWatcher(ctx context.Context) *clusterWatcher {
+	return &clusterWatcher{
+		ctx:      ctx,
+		mutex:    sync.RWMutex{},
+		handlers: newHandlerList(),
+		cancels:  newCancelSet(),
+	}
+}
+
+func (c *clusterWatcher) Run(local manager.Manager) error {
+	c.startManager(LocalCluster, local)
+	loop := controller.NewSecretReconcileLoop("cluster watcher", local)
+	return loop.RunSecretReconciler(c.ctx, c, kubeconfig.SecretPredicate{})
+}
+
+func (c *clusterWatcher) ReconcileSecret(obj *v1.Secret) (reconcile.Result, error) {
+	clusterName, cfg, err := kubeconfig.SecretToConfig(obj)
+	if err != nil {
+		return reconcile.Result{}, eris.Wrap(err, "failed to extract kubeconfig from secret")
+	}
+
+	// If the cluster already has a manager, remove the existing instance and start again.
+	if _, err := c.cancels.get(clusterName); err == nil {
+		c.removeCluster(clusterName)
+	}
+
+	mgr, err := manager.New(cfg.RestConfig, manager.Options{
+		// TODO these should be configurable, disable for now
+		MetricsBindAddress:     "0",
+		HealthProbeBindAddress: "0",
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	c.startManager(clusterName, mgr)
+
+	return reconcile.Result{}, nil
+}
+
+func (c *clusterWatcher) startManager(clusterName string, mgr manager.Manager) {
+	ctx, cancel := context.WithCancel(
+		contextutils.WithLoggerValues(context.WithValue(c.ctx, "cluster", clusterName), zap.String("cluster", clusterName)))
+	go func() {
+		err := mgr.Start(ctx.Done())
+		if err != nil {
+			contextutils.LoggerFrom(ctx).DPanicw("manager start failed for cluster %v", clusterName)
```

Comment:
Do we want to somehow propagate this error? We are using an `ErrorSignaler` for this in SMH
https://github.com/solo-io/service-mesh-hub/tree/master/pkg/errutils

Reply:
Can you point me to where the propagated errors are used? I like the idea but am reluctant to include an error signaler in this PR as it's already pretty large and I can't think of a way to programatically handle these errors beyond panicking and logging.

Reply:
In the [`AsyncManager`](https://github.com/solo-io/service-mesh-hub/blob/master/services/common/multicluster/manager/manager.go#L121). If an error occurs while starting the manager, it sends it to the signaler, which the caller can listen to.
Like so: https://github.com/solo-io/service-mesh-hub/blob/master/services/common/multicluster/common.go#L58

Reply:
Right, but we ultimately fatal if an error occurs: https://github.com/solo-io/service-mesh-hub/blob/2278e78e628302ac3e7fe5858254da4fc31f4848/services/mesh-discovery/run.go#L55.

I'd prefer to keep things simple for now and add a signaler later if there's more that we can do with the error received.